contributors/design-proposals/troubleshoot-running-pods.md

```diff
@@ -0,0 +1,666 @@
+# Troubleshoot Running Pods
+
+This proposal seeks to add first class support for troubleshooting by creating a
+mechanism to execute a shell or other troubleshooting tools inside a running pod
+without requiring that the associated container images include such tools.
+
+## Motivation
+
+### Development
+
+Many developers of native Kubernetes applications wish to treat Kubernetes as an
+execution platform for custom binaries produced by a build system. These users
+can forgo the scripted OS install of traditional Dockerfiles and instead `COPY`
+the output of their build system into a container image built `FROM scratch`.
+This confers several advantages:
+
+1.  **Minimal images** lower operational burden and reduce attack vectors.
+1.  **Immutable images** improve correctness and reliability.
+1.  **Smaller image size** reduces resource usage and speeds deployments.
+
+The disadvantage of using containers built `FROM scratch` is the lack of system
+binaries provided a base Operation System image makes it difficult to
+troubleshoot running containers. Kubernetes should enable troubleshooting pods
+regardless of the contents of the container image.
+
+### Operations and Support
+
+As Kubernetes gains in popularity, it's becoming the case that a person
+troubleshooting an application is not necessarily the person who built it.
+Operations staff and Support organizations want the ability to attach a "known
+good" or automated debugging environment to a pod.
+
+## Requirements
+
+A solution to troubleshoot arbitrary container images MUST:
+
+*   troubleshoot arbitrary running containers with minimal prior configuration
+*   allow access to shared pod namespaces and the file systems of individual
+    containers
+*   fetch troubleshooting utilities at debug time rather than at the time of pod
```

Comment:
@sjenning that's a clever solution that hadn't occurred to me, but I think it's solving a different problem. That sounds like a solution to a development or platform integration problem. Besides, let's not drop requirements to just make implementation easier. Let's solve the harder problem and delight the user.

Debug containers is an attempt to solve an operations and security problem. The operations problem is that I want to troubleshoot a running minimal container image (built from SCRATCH) without restarting it or having had to predict in advance which tools I would need.

Adding the container at pod creation has the following downsides:

- One of the motivations for minimal container images is to avoid deploying a full OS stack with every application instance to reduce attack surface.
- Specifying the debugging image at debug time allows freedom of choice between arbitrary debugging solutions. For example, I could choice between a "take snapshot and restart" container image or one that gives me an interactive shell.
- Choice of debugging image enables more useful cluster policy. For example, the "take snapshot and restart" image doesn't generate an audit alert while an interactive shell requires a human to review.
- Due to constraints of the runtime, a debugging container is only useful when it's running. Running an idle shell in hundreds of thousands of pods is inefficient and, frankly, feels like a hack.
- I want the latest debug tools rather than the old ones from when this pod was created.

These things are already mentioned in the proposal, but I understand that the proposal is too large for people to read because I've integrated hundreds (160 at last count, but that was before this PR was opened) of comments and suggestion.

Clearly I could be better at organizing information for this audience, and I would certainly appreciate any suggestions to that effect.